File: 0143-reorder-list/0143-reorder-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {void} Do not return anything, modify head in-place instead.
 */
var reorderList = function(head) {
    let fast = head;
    let slow = head;
    while (fast && fast.next && fast.next.next) {
        fast = fast.next.next;
        slow = slow.next;
    }
    let mid = slow.next;
    slow.next = null;
    let cnode = mid;
    let prev = null;
    while (cnode) {
        const temp = cnode.next;
        cnode.next = prev;
        prev = cnode;
        cnode = temp;
    }
    while (prev) {
        const temp = head.next;
        head.next = prev;
        prev = prev.next;
        head = head.next;
        head.next = temp;
        head = head.next;
    }
};

File: 0143-reorder-list/0143-reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        arr = []
        temp = head.next
        while temp:
            arr.append(temp.val)
            temp = temp.next
        temp = head.next
        i = 0
        j = len(arr) - 1
        while temp:
            temp.val = arr[j]
            j -= 1
            temp = temp.next
            if temp:
                temp.val = arr[i]
                i += 1
                temp = temp.next
        
        return head